reconcile/dynatrace_token_provider.py

```diff
@@ -0,0 +1,404 @@
+        # https://github.com/app-sre/qontract-schemas/pull/495/files is merged, need to update here to graphQL query
+        return ["ddl70254"]
+
+    def get_dynatrace_clients(
+        self, secret_reader: SecretReaderBase
+    ) -> Mapping[str, Dynatrace]:
+        tenant_ids = self.get_all_dynatrace_tenant_ids()
+        dynatrace_clients = {}
+        for tenant_id in tenant_ids:
+            # TODO: Below should be replaced by the query mentioned above
+            dt_bootstrap_token = secret_reader.read(
+                {
+                    "path": f"app-sre/creds/dynatrace/redhat-aws/bootstrap-api-tokens/{tenant_id}",
+                    "field": "token",
+                }
+            )
+            dt_client = Dynatrace(
+                f"https://{tenant_id}.live.dynatrace.com/",
+                dt_bootstrap_token,
+            )
+            dynatrace_clients[tenant_id] = dt_client
+        return dynatrace_clients
+
+    def process_cluster(
+        self,
+        dry_run: bool,
+        cluster: ClusterDetails,
+        dt_client: Dynatrace,
+        ocm_client: OCMBaseClient,
+    ) -> None:
+        existing_syncset = self.get_syncset(ocm_client, cluster)
+        if not existing_syncset:
+            if not dry_run:
+                try:
+                    (ingestion_token, operator_token) = self.create_dynatrace_tokens(
+                        dt_client, cluster.ocm_cluster.external_id
+                    )
+                    create_syncset(
+                        ocm_client,
+                        cluster.ocm_cluster.id,
+                        self.construct_syncset(ingestion_token, operator_token),
+                    )
+                except Exception as e:
+                    _expose_errors_as_service_log(
+                        ocm_client,
+                        cluster.ocm_cluster.external_id,
+                        f"DTP can't create Syncset with the tokens {str(e.args)}",
+                    )
+            logging.info(
+                f"Ingestion and operator tokens created in Dynatrace for cluster {cluster.ocm_cluster.external_id}."
+            )
+            logging.info(
+                f"SyncSet {SYNCSET_ID} created in cluster {cluster.ocm_cluster.external_id}."
+            )
+        else:
+            tokens = self.get_tokens_from_cluster(existing_syncset)
+            need_patching = False
+            for token_name, token in tokens.items():
+                if not self.token_exist_in_dynatrace(dt_client, token["id"]):
+                    need_patching = True
+                    logging.info(f"{token_name} missing in Dynatrace.")
+                    if token_name == DYNATRACE_INGESTION_TOKEN_NAME:
+                        if not dry_run:
+                            ingestion_token = self.create_dynatrace_ingestion_token(
+                                dt_client, cluster.ocm_cluster.external_id
+                            )
```

Comment:
what happens with orphaned tokens (token generated here but for whatever reason they did not make it into a syncset, e.g. because of a process crash, network issue, OCM issue, ...). would that be problematic? would we have a way to identify them clean them up later on?

i'm not suggesting that this needs to be solved right away. but maybe we can think this through, e.g. having a way to identify all tokens created through this integration so we can crosscheck later on which ones are used in syncsets and which ones are not.

Reply:
Ah yeah good point. The very least we can prefix those tokens with `dtp-` so they are easily recognizable. What do you think?

Reply:
good idea!